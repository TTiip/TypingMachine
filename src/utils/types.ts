interface InsertPatch {
  type: 'insert'
  from: number
  text: string
}

interface RemovePatch {
  type: 'remove'
  from: number
  length: number
}

type Patch = InsertPatch | RemovePatch

export {
  InsertPatch,
  RemovePatch,
  Patch
}
