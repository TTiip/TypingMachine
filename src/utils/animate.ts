import type { Patch } from '~/utils'

function* createAnimate (inputStr: string, patches: Patch[]) {
  let outPutStr = inputStr
  let cursor = 0
  for (const patch of patches) {
    if (patch.type === 'insert') {
      cursor = patch.from
      const head = outPutStr.slice(0, patch.from)
      const tail = outPutStr.slice(patch.from)
      let selection = ''

      for (const char of patch.text) {
        selection += char
        yield {
          cursor: cursor + selection.length,
          outPutStr: head + patch.text + tail
        }
      }
    } else if (patch.type === 'remove') {
      cursor = patch.from - patch.length
      const head = outPutStr.slice(0, patch.from - patch.length)
      const tail = outPutStr.slice(patch.from)
      let selection = outPutStr.slice(patch.from - patch.length, patch.from)

      for (let i = selection.length - 1; i >= 0; i--) {
        yield {
          cursor: cursor + i,
          outPutStr: head + selection.slice(0, i) + tail
        }
      }
    }
  }
  return outPutStr
}

export {
  createAnimate
}
