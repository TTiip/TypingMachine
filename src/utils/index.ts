export * from './patch'
export * from './typing'
export * from './animate'
export * from './types'
