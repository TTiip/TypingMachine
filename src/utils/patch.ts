import type { Diff } from 'diff-match-patch'
import { diff_match_patch as DMP } from 'diff-match-patch'
import type { Patch } from './types'

// 获取比较处理以后的片段
function diff (inputStr: string, outputStr: string): Diff[] {
  const differ = new DMP()
  const diff_main = differ.diff_main(inputStr, outputStr)
  differ.diff_cleanupEfficiency(diff_main)
  return diff_main
}

// 计算节点相关的东西
function calculatePatch (diffs: Diff[]): Patch[] {
  const patches: Patch[] = []
  let index = 0
  for (const diff of diffs) {
    if (diff[0] === 0) {
      index += diff[1].length
      continue
    }
    if (diff[0] === -1) {
      patches.push({
        type: 'remove',
        from: index + diff[1].length,
        length: diff[1].length
      })
    }
    if (diff[0] === 1) {
      patches.push({
        type: 'insert',
        from: index,
        text: diff[1]
      })
      index += diff[1].length
    }
  }
  return patches
}

function applyPatch (inputStr: string, patches: Patch[]): string {
  let outPutStr = inputStr
  for (const patch of patches) {
    if (patch.type === 'insert') {
      outPutStr = outPutStr.slice(0, patch.from) + patch.text + outPutStr.slice(patch.from)
    } else if (patch.type === 'remove') {
      outPutStr = outPutStr.slice(0, patch.from - patch.length) + outPutStr.slice(patch.from)
    }
  }
  return outPutStr
}

export {
  diff,
  calculatePatch,
  applyPatch
}
