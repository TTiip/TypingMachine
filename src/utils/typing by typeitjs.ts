import { createAnimate } from './animate'
import { calculatePatch, diff } from './patch'

const typingEl = document.querySelector('#typing') as HTMLTextAreaElement

function sleep (ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms))
}

async function start (inputStr: string, outputStr: string) {
  let _inputStr = inputStr
  const patches = calculatePatch(diff(_inputStr, outputStr))
  const animate = createAnimate(_inputStr, patches)

  for (const result of animate) {
    typingEl.textContent = result.outPutStr
    await sleep(Math.random() * 100)
  }
}
export {
  start
}
