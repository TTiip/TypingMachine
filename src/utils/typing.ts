import _TypeIt from 'typeit'
import { calculatePatch, diff } from './patch'

const TypeIt = _TypeIt as any
function start (inputStr: string, outputStr: string) {
  const patches = calculatePatch(diff(inputStr, outputStr))
  const typingEl = document.querySelector('#typing') as HTMLTextAreaElement

  // typingEl!.textContent = inputStr
  // typingEl!.textContent = ''

  // 创造实例 并保存
  let typeit: any = new TypeIt(typingEl, {
    speed: 100,
    waitUntilVisible: true
  })
    // 初始化 显示文本
    .type(inputStr, {
      instant: true,
      delay: 300
    })

  for (const patch of patches) {
    // 每个任务开始之前暂停 200ms
    typeit
      .pause(200)

    if (patch.type === 'insert') {
      // 先将光标移动到开始的位置，在讲光标移动到删除的位置。
      // 目前 分析是只能得出相对开始的位置。
      // instant 表示立即放置光标到指定位置
      // delay 表示完成之后延迟多久 执行下一步
      typeit
        .move(null, {
          to: 'START',
          instant: true
        })
        .move(patch.from, { instant: true })
        .type(patch.text, { delay: 500 })
    } else if (patch.type === 'remove') {
      // 逻辑同上。
      typeit
        .move(null, {
          to: 'START',
          instant: true
        })
        .move(patch.from, { instant: true })
        .delete(patch.length)
    }
  }
  // 运行 打字机
  typeit.go()
}
export {
  start
}
