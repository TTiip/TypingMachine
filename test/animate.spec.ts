import { describe, expect, it } from 'vitest'
import { inputStr, outputStr } from './fixture'
import { calculatePatch, createAnimate, diff } from '~/utils'

describe('animate', () => {
  it('animate', () => {
    const delta = diff(inputStr, outputStr)
    const patches = calculatePatch(delta)
    const animate = createAnimate(inputStr, patches)

    expect([...animate]).toMatchSnapshot()
  })
})
