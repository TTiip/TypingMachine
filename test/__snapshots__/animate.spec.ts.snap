// Vitest Snapshot v1

exports[`animate > animate 1`] = `
[
  {
    "cursor": 74,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 73,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 72,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './sr
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 71,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './s
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 70,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 69,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from '.
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 68,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from '
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 67,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from 
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 66,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 65,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } fro
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 64,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } fr
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 63,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } f
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 62,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } 
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 61,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one }
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 60,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one 
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 59,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 58,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { on
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 57,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { o
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 56,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { 
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 55,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import {
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 54,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import 
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 53,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 52,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
impor
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 51,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
impo
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 50,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
imp
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 49,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
im
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 48,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
i
describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 47,
    "outPutStr": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 88,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('shuld', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 87,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('suld', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 86,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('uld', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 85,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe(uld', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 84,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describeuld', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 83,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describuld', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 82,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

descriuld', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 81,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

descruld', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 82,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

descroneibe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 83,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

descroneibe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 84,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

descroneibe('should', () => {
  it('exported', () => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 121,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 122,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 123,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 124,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 125,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 126,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 127,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 128,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 129,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 130,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 131,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 132,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 133,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 134,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 135,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 136,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 137,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 138,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 139,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 140,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 141,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 142,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 143,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 144,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 145,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 146,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
  {
    "cursor": 147,
    "outPutStr": "
import { describe, expect, it } from 'vitest'
import { one } from './src'

describe('should', () => {
  it('exported', )
    expect(two).toEqual(2() => {
    expect(one).toEqual(1)
  })
})
",
  },
]
`;
